Code42/allow null secure random number gen

EricChristensen opened the pull request:
This pull request allows the GoogleAuthenticator class to be initialized with user specified values for the secure random algorithm and secure random algorithm provider. This is necessary for us because we will not necessarily have the default algorithm and algorithm provider currently hard coded in the project.

emcrisostomo commented:
Thanks @EricChristensen, I'll have a look at it.

ZuluForce commented on src/main/java/com/warrenstrange/googleauth/GoogleAuthenticator.java:
```diff
@@ -166,10 +168,56 @@ public GoogleAuthenticator(GoogleAuthenticatorConfig config)
+
+    public GoogleAuthenticator(final String randomNumberAlgorithm, final String randomNumberAlgorithmProvider)
+    {
+        config = new GoogleAuthenticatorConfig();
```
Should this constructor chain to the one below to avoid duplicating the preconditions?

EricChristensen replied:
Good idea https://github.com/wstrange/GoogleAuth/pull/87/commits/0459feb0190bb7e9d4ee6959a2c9a10f5e9963c0

EricChristensen commented:
@emcrisostomo Do you know when you'll take a look at this? There is no time pressure because we have a temporary solution, my manager just asked if I knew if we could expect a time frame

emcrisostomo commented:
I'll to have a look at it this week

jdullum commented on src/main/java/com/warrenstrange/googleauth/GoogleAuthenticator.java:
```diff
@@ -166,10 +168,44 @@ public GoogleAuthenticator(GoogleAuthenticatorConfig config)
+
+    }
+
+    public GoogleAuthenticator(GoogleAuthenticatorConfig config, final String randomNumberAlgorithm, final String randomNumberAlgorithmProvider)
```
Did you consider adding `randomNumberAlgorithm` and `randomNumberAlgorithmProvider` to the `GoogleAuthenticatorConfig`? I'm not sure why these particular properties were only exposed via system properties. But if we're going to expose another way to configure them, the config might be better than an overloaded constructor.

EricChristensen replied:
@emcrisostomo I thought about adding the change that @jdullum is referring to. I think that that would be better than the current change I made. My hesitation is that `DEFAULT_RANDOM_NUMBER_ALGORITHM` and `DEFAULT_RANDOM_NUMBER_ALGORITHM_PROVIDER` were already defined in GoogleAuthenticator. Adding the algorithm and provider to the config would mean moving the default algorithm and provider to the config, which would be a refactor. Since this is your repo and not ours I want to follow existing patterns and be minimally invasive. That being said, what are your thoughts on the change described? I'd be happy to make that change or stick to the existing pattern.

emcrisostomo replied:
I agree with you @jdullum , @EricChristensen , I agree with this change, I'm going to merge this anyway, and if you're willing to provide a PR I'll be happy to consider it.

jdullum commented on src/main/java/com/warrenstrange/googleauth/GoogleAuthenticator.java:
```diff
@@ -166,10 +168,44 @@ public GoogleAuthenticator(GoogleAuthenticatorConfig config)
+
+        this.config = config;
+
+        if (randomNumberAlgorithm == null && randomNumberAlgorithmProvider == null)
```
It would be nice to update the javadoc for these parameters, especially since they allow null and don't use `DEFAULT_RANDOM_NUMBER_ALGORITHM` or `DEFAULT_RANDOM_NUMBER_ALGORITHM_PROVIDER `.

EricChristensen replied:
https://github.com/wstrange/GoogleAuth/pull/87/commits/98274b6e6b8835353b00ee93d2f0c174798a1035

EricChristensen commented:
@emcrisostomo any thoughts? We may move forward with forking this repo and going with that

EricChristensen commented:
> @EricChristensen, I'm afraid I'm busy on other things at the moment and I haven't looked into it thoroughly. Also, as @jdullum has mentioned, the library is documented, so a pull request that doesn't update the relevant documentation (both the JavaDoc _and_ the texinfo doc) is incomplete. In any case, I'll have a look at it as soon as I can.

Update to the Javadoc and the texinfo doc https://github.com/wstrange/GoogleAuth/pull/87/commits/98274b6e6b8835353b00ee93d2f0c174798a1035

EricChristensen commented:
@emcrisostomo Don't know if you saw, but I addressed your comment about the documentation here https://github.com/wstrange/GoogleAuth/pull/87/commits/98274b6e6b8835353b00ee93d2f0c174798a1035

EricChristensen commented:
@emcrisostomo Did you see my previous comment?

emcrisostomo commented:
Yes, apologies, I've seen it but had no time to address this PR yet.

On Tue, Jun 2, 2020 at 4:37 PM Eric Christensen <notifications@github.com>
wrote:

> @emcrisostomo <https://github.com/emcrisostomo> Did you see my previous
> comment?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wstrange/GoogleAuth/pull/87#issuecomment-637585054>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSWO2VASHDU5GLLGEV2IQLRUUFD7ANCNFSM4K7WKJEA>
> .
>

wstrange commented:
@emcrisostomo Good to merge?
